### ISSUE TITLE
CQ issues with HW intrinsics that return bool

### ISSUE THREAD

**Issue by mikedn** [dotnet/coreclr#17073]
HW intrinsics that produce a bool value need special handling when the value they produce is used by a conditional branch. As is now the intrinsic node produces the value in a register and then an additional test is required to check for false/true and branch.
For example
```C#
return Sse2.CompareEqualOrderedScalar(x, y) ? 4 : 2;
```
generates
```asm
       C4E1792FC1           vcomisd  xmm0, xmm1
       0F9BC2               setpo    dl
       0F94C0               sete     al
       22C2                 and      al, dl
       0FB6D0               movzx    rdx, al
       85D2                 test     edx, edx
       7506                 jne      SHORT G_M55886_IG04
       B802000000           mov      eax, 2
G_M55886_IG03:
       C3                   ret
G_M55886_IG04:
       B804000000           mov      eax, 4
G_M55886_IG05:
       C3                   ret
```
but it should generate something like
```asm
G_M55886_IG02:
       C4E1792EC1           vcomisd xmm0, xmm1
       7A02                 jpe      SHORT G_M55886_IG03
       7406                 je       SHORT G_M55886_IG05
G_M55886_IG03:
       B802000000           mov      eax, 2
G_M55886_IG04:
       C3                   ret
G_M55886_IG05:
       B804000000           mov      eax, 4
G_M55886_IG06:
       C3                   ret
```
As far as I can tell this affects pretty much all HW intrinsics that return bool. Except `IsSupported`, of course.

category:cq
theme:intrinsics
skill-level:expert
cost:medium

**Comment by mikedn** [comment]
Fixing this will probably require handling such nodes in lowering, similar to what we do for SIMD equality/inequality in `LowerSIMD`. Though it's going to be a bit more complicated due to the need to support floating point conditions...

I can probably take care of this post 2.1.

**Comment by RussKeldorph** [comment]
@fiigii @dotnet/jit-contrib

**Comment by fiigii** [comment]
> Fixing this will probably require handling such nodes in lowering, similar to what we do for SIMD equality/inequality in LowerSIMD. 

@mikedn Thank you so much for opening this issue and providing the solution. I have also considered this problem. Looks like we can use `GTF_USE_FLAGS` to guide the backend generating the "bool value" or not, right?

**Comment by mikedn** [comment]
> Looks like we can use GTF_USE_FLAGS to guide the backend generating the "bool value" or not, right?

Probably not, GTF_USE_FLAGS is mostly useless (and confusing). `LowerSIMD` changes the type of the SIMD node to `TYP_VOID` and then relies on SETCC to produce a bool value or JCC to do a conditional branch. In some other cases we only change the node type to `TYP_VOID` only when the node is used together with a conditional branch and leave the type unchanged otherwise (and expect codegen to produce the bool value). We'll see.

Right now the main problem is dealing with floating point conditions as neither SETCC nor JCC support them. Send my warm regards to whoever designed floating point compares at Intel, they messed up :grin:

**Comment by briansull** [comment]
The Intel 8087, announced in 1980, was the first x87 floating-point coprocessor for the 8086 line of microprocessors.
The 8087 was initially conceived by Bill Pohlman, the engineering manager at Intel who oversaw the development of the 8086 chip. Bill took steps to be sure that the 8086 chip could support a yet-to-be-developed math chip.
The 8087 was an advanced IC for its time, pushing the limits of period manufacturing technology. Initial yields were extremely low.
https://en.wikipedia.org/wiki/Intel_8087

**Comment by CarolEidt** [comment]
I agree that the right approach would be similar to what's done for the SIMD equality comparisons, and that the trickiest part will be adding support for floating point conditions. Do you think you would take the approach of extending `JCC` and `SETCC` or distinct nodes for floating point conditions?

**Comment by mikedn** [comment]
> Do you think you would take the approach of extending JCC and SETCC or distinct nodes for floating point conditions?

Likely, I have code for that in the if-conversion pull request. The main problem I have is with FP condition reversal in an attempt to use FP conditions that produce simpler code. 
* For relop + JTRUE reversal is done in `genCompareFloat` and `ContainCheckCompare`
* For CMP + JCC reversal has to be done in lowering because you need to know the condition to swap compare operands but since the condition moves to JCC you can no longer swap compare operands in codegen

Now, it's IMO more natural to do it in lowering rather than doing it in both contain check and codegen. But we run liveness and update flow graph after lowering and the later may reverse branches as a result of DCE. It will work correctly but it will undo the FP condition selection lowering did in an attempt to produce simpler code.  We'll see, I kind of hope that such cases will be rare.

**Comment by mikedn** [comment]
Expanded JCC condition support available in #17733
